org.lflang/src/org/lflang/generator/ts/TSReactorGenerator.kt

```diff
@@ -0,0 +1,715 @@
+    }
+    private fun initializeParameter(p: Parameter): String {
+        return """${p.name}: ${getTargetType(p)} = ${getTargetInitializer(p)}"""
+    }
+
+    /**
+     * Return a TS type for the specified action.
+     * If the type has not been specified, return
+     * "Present" which is the base type for Actions.
+     * @param action The action
+     * @return The TS type.
+     */
+    private fun getActionType(action: Action): String {
+        if (action.type !== null) {
+            return getTargetType(action.type)
+        } else {
+            return "Present"
+        }
+    }
+
+    /**
+     * Return a TS type for the specified port.
+     * If the type has not been specified, return
+     * "Present" which is the base type for ports.
+     * @param port The port
+     * @return The TS type.
+     */
+    private fun getPortType(port: Port): String {
+        if (port.type !== null) {
+            return getTargetType(port.type)
+        } else {
+            return "Present"
+        }
+    }
+
+    private fun generateArg(v: VarRef): String {
+        return if (v.container !== null) {
+            """__${v.container.name}_${v.variable.name}"""
+        } else {
+            """__${v.variable.name}"""
+        }
+    }
+
+    // TODO(hokeun): Split this method into smaller methods.
+    fun generateReactorFederated(reactor: Reactor, federate: FederateInstance) {
+        pr("// =============== START reactor class " + reactor.name)
+
+        for (p in reactor.preambles?: emptyList()) {
+            pr("// *********** From the preamble, verbatim:")
+            pr(p.code.toText())
+            pr("\n// *********** End of preamble.")
+        }
+
+        var reactorName = reactor.name
+        if (!reactor.typeParms.isEmpty()) {
+            reactorName +=
+                reactor.typeParms.joinToString(", ", "<", ">") { it.toText() }
+        }
+        // NOTE: type parameters that are referenced in ports or actions must extend
+        // Present in order for the program to type check.
+        if (reactor.isMain()) {
+            pr("class " + reactorName + " extends __App {")
+        } else if (reactor.isFederated()) {
+            pr("class " + reactorName + " extends __FederatedApp {")
+        } else {
+            pr("export class " + reactorName + " extends __Reactor {")
+        }
+
+        indent()
+
+        var arguments = LinkedList<String>()
+        if (reactor.isMain() || reactor.isFederated()) {
+            arguments.add("timeout: TimeValue | undefined = undefined")
+            arguments.add("keepAlive: boolean = false")
+            arguments.add("fast: boolean = false")
+        } else {
+            arguments.add("parent: __Reactor")
+        }
+
+        // For TS, parameters are arguments of the class constructor.
+        for (parameter in reactor.parameters) {
+            arguments.add(initializeParameter(parameter))
+        }
+
+        var reactorConstructor = StringBuilder()
```

Comment:
Nearly all of the `var` in this file could be `val` instead

Reply:
Done.